src/core/unit-dependency-atom.c

```diff
@@ -99,13 +99,13 @@ static const UnitDependencyAtom atom_map[_UNIT_DEPENDENCY_MAX] = {
         [UNIT_REFERENCED_BY]          = UNIT_ATOM_REFERENCED_BY,
         [UNIT_IN_SLICE]               = UNIT_ATOM_IN_SLICE,
         [UNIT_SLICE_OF]               = UNIT_ATOM_SLICE_OF,
+        [UNIT_ON_SUCCESS_OF]          = UNIT_ATOM_PROPAGATE_EXIT_STATUS,
+        [UNIT_ON_FAILURE_OF]          = UNIT_ATOM_PROPAGATE_EXIT_STATUS,
```

Comment:
i think these probably should also imply referencing, i.e. shoudl also have te UNIT_REFERENCES atom. So that a failure/success handler will always pin the unit is triggered by, so that it can collect the exit data reliably.

(And the UNIT_ATOM_BACK_REFERENCE_IMPLIED atom should go away)